### PR TITLE
TMEDIA-573 - Account Information - Password

### PR DESCRIPTION
**Pull request by matthewroach**
## Description

Building upon the working in #1181 

Adding password feature to account information

## Jira Ticket
- [TMEDIA-573](https://arcpublishing.atlassian.net/browse/TMEDIA-573)

## Acceptance Criteria

_See ticket_

## Test Steps
- Add test steps a reviewer must complete to test this PR

1. Checkout this branch `git checkout TMEDIA-573-account-information-password`
2. Run fusion repo with linked blocks `npx fusion start -f -l @wpmedia/identity-block`
3. Create an Account Management Page or use the one created in #1181 
4. If Account Management block is on page, remove it and re-add it
5. Validate there is now an additional custom field - Enable Password Editing
6. Publish Page
7. Create an account using Email and Password - http://localhost/account/sign-up/?_website=the-gazette
8. Go to the Account Management Page
9. Change your password
10. Logout
11. Login using new password

## Effect Of Changes

Microsoft Edge-2021-12-03-13-38 33

## Dependencies or Side Effects
Requires #1181 to be merged first

## Author Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [x] Confirmed this PR has unit test files
  - [x] Ran `npm run test`, made sure all tests are passing
  - [x] If the amount of work to write unit tests for this change are excessive,
please explain why (so that we can fix it whenever it gets refactored).
- [x] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist
_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] Linting code actions have passed.
- [ ] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr.
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.

### CODE REVIEW

File: blocks/identity-block/features/account-management/default.jsx [hunk 1]
```diff
@@ -82,10 +85,11 @@ function AccountManagement({ customFields }) {
...
     <AccountManagementPresentational header={header}>
       {
-        showEmail && (
-          <EmailEditableFieldContainer email={email} onSubmit={handleEmailUpdate} />
-        )
+        showEmail ? (
+          <EmailEditableFieldContainer email={email} />
+        ) : null
       }
+      {showPassword ? <PasswordEditableFieldContainer email={email} hasPassword={hasPassword} /> : null}
```
Comment (JackHowa) [inline blocks/identity-block/features/account-management/default.jsx]: passing in email to PassWordEditableFieldContainer?

File: blocks/identity-block/intl.json [hunk 2]
```diff
@@ -262,5 +262,93 @@
...
+		"es": "Save",
+		"fr": "Save",
+		"ja": "Save",
+		"ko": "Save",
+		"no": "Save",
+		"pt": "Save",
+		"sv": "Save"
+	},
+	"identity-block.cancel": {
+		"de": "Cancel",
```
Comment (JackHowa) [inline blocks/identity-block/intl.json]: nice -- thank you

File: blocks/identity-block/components/EditableFormInputField/index.jsx [hunk 3]
```diff
@@ -74,51 +66,57 @@ function EditableFieldPresentational({
...
                     !isEditable
                       && (
-                        <ButtonLink
+                        <PrimaryFont
+                          as="button"
                           className="editable-form-input--edit-button-link"
                           type="button"
                           onClick={() => setIsEditable(true)}
-                          color={primaryColor}
+                          fontColor="primary-color"
```
Comment (JackHowa) [inline blocks/identity-block/components/EditableFormInputField/index.jsx]: nice -- forgot this one

File: blocks/identity-block/features/account-management/_children/EmailEditableFieldContainer.jsx [hunk 4]
```diff
@@ -1,30 +1,52 @@
...
+    });
+
+  const handleCancelEdit = () => {
+    setError(null);
+  };
 
   return (
     <EditableFieldPresentational
-      initialValue={email}
+      initialValue={emailAddress || email}
```
Comment (JackHowa) [inline blocks/identity-block/features/account-management/_children/EmailEditableFieldContainer.jsx]: can we name these like incomingEmail or current email? email and emailAddress doesn't really help me understand it; likely just a merge issue

File: blocks/identity-block/features/account-management/_children/EmailEditableFieldContainer.jsx [hunk 5]
```diff
@@ -1,30 +1,52 @@
...
       label={formEmailLabel}
       editText={editText}
-      onSubmit={onSubmit}
+      onSubmit={handleEmailUpdate}
+      formErrorText={error}
```
Comment (JackHowa) [inline blocks/identity-block/features/account-management/_children/EmailEditableFieldContainer.jsx]: nice -- thank you

File: blocks/identity-block/features/account-management/_children/EmailEditableFieldContainer.jsx [hunk 6]
```diff
@@ -1,30 +1,52 @@
...
+      cancelEdit={handleCancelEdit}
     >
       <FormInputField
         type={FIELD_TYPES.EMAIL}
         label={formEmailLabel}
-        defaultValue={email}
+        defaultValue={emailAddress || email}
```
Comment (JackHowa) [inline blocks/identity-block/features/account-management/_children/EmailEditableFieldContainer.jsx]: same thing here as above

File: blocks/identity-block/features/account-management/_children/PasswordEditableFieldContainer.jsx [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+import getProperties from 'fusion:properties';
+import getTranslatedPhrases from 'fusion:intl';
+import { useIdentity } from '../../..';
+import EditableFieldPresentational from '../../../components/EditableFormInputField';
+import FormInputField, { FIELD_TYPES } from '../../../components/FormInputField';
+import FormPasswordConfirm from '../../../components/FormPasswordConfirm';
+import passwordValidationMessage from '../../../utils/password-validation-message';
+import validatePasswordPattern from '../../../utils/validate-password-pattern';
+
+function PasswordEditableFieldContainer({ email, hasPassword }) {
```
Comment (JackHowa) [inline blocks/identity-block/features/account-management/_children/PasswordEditableFieldContainer.jsx]: don't need email here right?

File: blocks/identity-block/features/account-management/_children/PasswordEditableFieldContainer.jsx [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    pwPwNumbers = 0,
+    pwSpecialCharacters = 0,
+    pwUppercase = 0,
+    status,
+  } = passwordRequirements;
+
+  const passwordErrorMessage = passwordValidationMessage({
+    defaultMessage: phrases.t('identity-block.password-requirements'),
+    options: {
+      lowercase: {
```
Comment (JackHowa) [inline blocks/identity-block/features/account-management/_children/PasswordEditableFieldContainer.jsx:47]: wonder if there's a way to extract these requirements. these are from another place in the repo, right?

File: blocks/identity-block/features/account-management/_children/PasswordEditableFieldContainer.jsx [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+      cancelEdit={handleCancelEdit}
+    >
+      {hasPassword ? (
+        <>
+          <FormInputField
+            type={FIELD_TYPES.PASSWORD}
+            label={phrases.t('identity-block.current-password')}
+            showDefaultError={false}
+            required
+            autoComplete="current-password"
```
Comment (JackHowa) [inline blocks/identity-block/features/account-management/_children/PasswordEditableFieldContainer.jsx:118]: nice

File: blocks/identity-block/features/account-management/default.jsx [hunk 10]
```diff
@@ -4,22 +4,27 @@ import PropTypes from '@arc-fusion/prop-types';
...
+import PasswordEditableFieldContainer from './_children/PasswordEditableFieldContainer';
+
+import './styles.scss';
 
 export function AccountManagementPresentational({ header, children }) {
   return (
-    <div>
-      <h1>{header}</h1>
+    <div className="account-management-layout">
+      <PrimaryFont as="h2">{header}</PrimaryFont>
```
Comment (JackHowa) [inline blocks/identity-block/features/account-management/default.jsx]: nice

File: blocks/identity-block/features/account-management/default.jsx [hunk 11]
```diff
@@ -51,11 +50,15 @@ function AccountManagement({ customFields }) {
...
       .getUserProfile()
       .then((profileObject) => {
-        const { email: loggedInEmail } = profileObject;
+        const { email: loggedInEmail, identities } = profileObject;
 
         if (loggedInEmail) {
           setEmail(loggedInEmail);
         }
+
+        const passwordProfile = identities.filter(({ type }) => type === 'Password' || type === 'Identity');
```
Comment (JackHowa) [inline blocks/identity-block/features/account-management/default.jsx:62]: good work

**Review by JackHowa — COMMENTED** [review]
I see this is a draft. But I can definitely use some of this for my pr. let me know if the existing feedback makes sense

File: blocks/identity-block/components/EditableFormInputField/index.jsx [hunk 12]
```diff
@@ -58,10 +47,13 @@ function EditableFieldPresentational({
...
   editText,
   onSubmit,
+  cancelEdit,
+  formErrorText,
 }) {
+  const [isEditable, setIsEditable] = useState(!!formErrorText);
   const { arcSite } = useFusionContext();
-  const primaryColor = getThemeStyle(arcSite)['primary-color'];
-  const [isEditable, setIsEditable] = useState(false);
+  const { locale } = getProperties(arcSite);
```
Comment (JackHowa) [inline blocks/identity-block/components/EditableFormInputField/index.jsx]: I was trying to leave as much fusion hooks out of the components folder. I may try to pass in this phrases with the save text

File: blocks/identity-block/components/EditableFormInputField/index.test.jsx [hunk 13]
```diff
@@ -3,6 +3,15 @@ import { mount } from 'enzyme';
 
 import EditableFormInputField, { ConditionalFormContainer } from '.';
 
+jest.mock('fusion:properties', () => (jest.fn(() => ({
```
Comment (JackHowa) [inline blocks/identity-block/components/EditableFormInputField/index.test.jsx]: if we pass in save text as props, we wouldn't need this mock

File: blocks/identity-block/components/EditableFormInputField/index.jsx (hunk 12 above)
Comment (JackHowa) [reply blocks/identity-block/components/EditableFormInputField/index.jsx]: updating now

File: blocks/identity-block/features/account-management/default.jsx (hunk 1 above)
Comment (matthewroach) [reply blocks/identity-block/features/account-management/default.jsx]: Yes, as if they don't have a password set you need the email to create their password account

File: blocks/identity-block/features/account-management/_children/EmailEditableFieldContainer.jsx [hunk 14]
```diff
@@ -1,30 +1,52 @@
...
   const phrases = getTranslatedPhrases(locale);
 
+  const { Identity } = useIdentity();
+
   const formEmailLabel = phrases.t('identity-block.email');
   const emailRequirements = phrases.t('identity-block.email-requirements');
   const editText = phrases.t('identity-block.edit');
+  const emailError = phrases.t('identity-block.update-email-error');
+
+  const handleEmailUpdate = ({ email: newEmail }) => Identity
```
Comment (JackHowa) [inline blocks/identity-block/features/account-management/_children/EmailEditableFieldContainer.jsx]: todo jack: handle error text at this level

File: blocks/identity-block/features/account-management/_children/PasswordEditableFieldContainer.jsx [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+        .updatePassword(oldPassword, newPassword)
+        .then(() => {
+          setError(false);
+        }).catch(() => {
+          setError(phrases.t('identity-block.update-password-error'));
+          throw new Error();
+        });
+    }
+    return Identity.signUp(
+      { userName: email, credentials: newPassword },
```
Comment (JackHowa) [inline blocks/identity-block/features/account-management/_children/PasswordEditableFieldContainer.jsx]: you are using email here, didn't see that

Comment (JackHowa) [reply blocks/identity-block/features/account-management/_children/PasswordEditableFieldContainer.jsx]: if you sign in with google, then use a different email without updating the page. then try to create a password without updating the page. the email would need to be the updated email.

the create password flow from a google sign in with a different email would require a verified email. with an unverified one, this process with google sign in becomes unmanageable.

**Review by nschubach — APPROVED** [review]
Seems to do what it needs, code looks good too.

**Pull request merged**